WTF::map in NativePromise to build Vectors

jyavenard opened the pull request:
#### 0f9d5afd257ecb62c8e877f2df163c6140188d58
<pre>
WTF::map in NativePromise to build Vectors
<a href="https://bugs.webkit.org/show_bug.cgi?id=263176">https://bugs.webkit.org/show_bug.cgi?id=263176</a>
rdar://116991334

Reviewed by Chris Dumez.

No change in functional behaviour.

* Source/WTF/wtf/NativePromise.h:

Canonical link: <a href="https://commits.webkit.org/269382@main">https://commits.webkit.org/269382@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/923c48c059d9745d1cc92d3fe749902597d497d3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22323 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22521 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23398 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24223 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20659 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26783 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22852 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21668 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22559 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22255 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19374 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25078 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19305 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20229 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26478 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19413 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20322 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20460 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24339 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/21691 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20992 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17784 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/25742 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20449 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6060 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5335 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24680 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/27017 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20987 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5887 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -530,14 +530,11 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
-
-                m_producer->resolve(WTFMove(resolveValues));
-                m_producer = nullptr;
+                m_producer->resolve(Vector<ResolveValueType>(m_resolveValues.size(), [&](size_t index) {
```
Seems like a better use case for WTF::map()?

```
m_producer->resolve(WTF::map(WTFMove(m_resolveValues), [](auto&& resolveValue) -> ResolveValueType {
    return WTFMove(resolveValue.value());
});
```

cdumez commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -530,14 +530,10 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
+                    return WTFMove(*resolveValue);
+                }));
                 m_producer = nullptr;
-                m_resolveValues.clear();
```
Don't we still want to clear `m_resolveValues`? Otherwise, we should consider using `std::exchange(m_resolveValues, { })` instead of `WTFMove(m_resolveValues)` above.

cdumez commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -581,11 +576,9 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
-                Vector<Result> results(m_results.size(), [&](size_t index) {
-                    return WTFMove(*m_results[index]);
-                });
-                m_results.clear();
```
Same comment.

cdumez reviewed (approved):
Seems like the clearing of the vector is not strictly based on slack discussion so LGTM with or without vector clearing, as long as the bots are happy.

jyavenard committed:
WTF::map in NativePromise to build Vectors
https://bugs.webkit.org/show_bug.cgi?id=263176
rdar://116991334

Reviewed by Chris Dumez.

No change in functional behaviour.

* Source/WTF/wtf/NativePromise.h:

Canonical link: https://commits.webkit.org/269382@main
